src/harness/vfs.ts

```diff
@@ -671,6 +671,160 @@ namespace vfs {
+            while (!baseNode.links && baseNode.shadowRoot) baseNode = baseNode.shadowRoot;
+
+            // no difference if the nodes are the same reference
+            if (changedNode === baseNode) return false;
+
+            // no difference if both nodes are non shadowed and have no entries
+            if (isEmptyNonShadowedDirectory(changedNode) && isEmptyNonShadowedDirectory(baseNode)) return false;
+
+            // no difference if both nodes are unpopulated and point to the same mounted file system
+            if (!changedNode.links && !baseNode.links &&
+                changedNode.resolver && changedNode.source !== undefined &&
+                baseNode.resolver === changedNode.resolver && baseNode.source === changedNode.source) return false;
+
+            // no difference if both nodes have identical children
+            const children: FileSet = {};
+            if (!FileSystem.diffWorker(children, changed, changed._getLinks(changedNode), base, base._getLinks(baseNode))) {
+                return false;
+            }
+
+            container[basename] = new Directory(children);
+            return true;
+        }
+
+        private static fileDiff(container: FileSet, basename: string, changed: FileSystem, changedNode: FileInode, base: FileSystem, baseNode: FileInode) {
+            while (!changedNode.buffer && changedNode.shadowRoot) changedNode = changedNode.shadowRoot;
+            while (!baseNode.buffer && baseNode.shadowRoot) baseNode = baseNode.shadowRoot;
+
+            // no difference if the nodes are the same reference
+            if (changedNode === baseNode) return false;
+
+            // no difference if both nodes are non shadowed and have no entries
+            if (isEmptyNonShadowedFile(changedNode) && isEmptyNonShadowedFile(baseNode)) return false;
+
+            // no difference if both nodes are unpopulated and point to the same mounted file system
+            if (!changedNode.buffer && !baseNode.buffer &&
+                changedNode.resolver && changedNode.source !== undefined &&
+                baseNode.resolver === changedNode.resolver && baseNode.source === changedNode.source) return false;
+
+            const changedBuffer = changed._getBuffer(changedNode);
+            const baseBuffer = base._getBuffer(baseNode);
+
+            // no difference if both buffers are the same reference
+            if (changedBuffer === baseBuffer) return false;
+
+            // no difference if both buffers are identical
+            if (Buffer.compare(changedBuffer, baseBuffer) === 0) return false;
+
+            container[basename] = new File(changedBuffer);
+            return true;
+        }
+
+        private static symlinkDiff(container: FileSet, basename: string, changedNode: SymlinkInode, baseNode: SymlinkInode) {
+            // no difference if the nodes are the same reference
+            if (changedNode.symlink === baseNode.symlink) return false;

```

Comment:
If symlinks are interpreted as relative paths, should a symlink at `/lib/foo` of `../bar` be equivalent to one at the same location of `../../lib/bar`? Or am I mistaken?

Reply:
This compares symlinks in the same path, so you shouldn't run into this.